### ISSUE TITLE
175B opt model seems to output gibberish

### ISSUE THREAD

**Issue by zhisbug** [facebookresearch/metaseq#122]
#### What is your question?

We recently downloaded the open sourced OPT-175B weight and managed to set it up using our [own Alpa](https://github.com/alpa-projects/alpa) system, and tried to play with it. Thanks for the effort of making it public!

Unfortunately, we found that the175B model outputs gibberish. An example:

_Prompt: Computer science is the study of computation
Output: Computer science is the study of computation and a gotten hoped planned done hoped also to practice to to a come to become nothing to to good taken become hoped to placed originally better only to to just gotten better just given done issues hoped good come come some'_

The way we process the 175B model is as follows:
1. use the `scripts/reshard_mp.py` to merge FSDP shards (192) into one for each model-parallel part, resulting in 8 parts.
2. use the `scripts/consolidate_fsdp_shard.py` to merge the 8 parts into one consolidated version (the core function used here is `glue_megatron_parts()`)
3. then we load this model into our system.

All the other models processed following this procedure (from 125M to 30B ones) work pretty well, except that the 175B outputs gibberish.

We also cross-checked with HuggingFace's implementation models from 125M to 30B -- the output logits are the same on models. Hence it shouldn't be a problem of the Alpa system.

Here is our model spec: https://github.com/alpa-projects/alpa/blob/torch_transformer_model/playground/model/opt_model.py#L499

We're wondering if you could give any hints? Several questions:

1. we downloaded the 992 shards about 2 weeks ago. Since then, have you updated the weights?
2. is there any architectural changes specific to the 175B weight we're unaware of?

#### What's your environment?

 - metaseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.10
 - OS (e.g., Linux): Ubuntu
 - How you installed metaseq (`pip`, source): source
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: 32x V100
 - Any other relevant information:

**Comment by stephenroller** [comment]
Hm, can you try just consolidate FSDP directly from the MP shards?

The checkpoint hasn't changed. Our own interactive_hosted expected the MP version (your step1) if you want to compare.

**Comment by zhisbug** [comment]
@stephenroller Thanks for the response! What do you mean by "MP version"? is it the state after FSDP sharding being consolidated? If yes -- since you use MP = 8, I assume your internal script will start from 8 files (each file corresponds to an MP part), right?

**Comment by anj-s** [comment]
> @stephenroller Thanks for the response! What do you mean by "MP version"? is it the state after FSDP shards are merged? If yes -- since you use MP = 8, I assume your internal script will start from 8 files (each file corresponds to an MP part), right?

MP refers to the model/tensor parallel checkpoints. There are a version of checkpoints that are purely tensor parallel(MP) in format.

**Comment by stephenroller** [comment]
Can you also double check your md5sums? I've seen at least one person get gibberish when the consolidation script silently failed with some weird corruption we never understood.  Re-consolidating fixed it

**Comment by zhisbug** [comment]
@stephenroller we actually run the consolidation 3 times using three different methods:
(1) Use `reshard_mp.py` to merge 992 shards into 8 MP parts; then use `consolidate_fsdp_shard.py` to merge 8 parts into 1 consolidated checkpoint.
(2) Directly use `consolidate_fsdp_shard.py` to merge 992 shards into 1 checkpoint file (`reshard_mp.py` is not used here).
(3) Use `reshard_mp.py` to merge 992 shards into 8 MP parts, then use `covert_to_singleton.py` which inits the model in CPU memory and load the state dict, then use allgather the gather and merge weights.


We run `np.assert_allclose()` on (randomly sampled) weights of the consolidated file generated by all (1)(2)(3) -- all passed -- meaning the weights generated by any path is the same.

So it is unlikely caused by errors/bugs in the consolidation script.

**Comment by stephenroller** [comment]
I'm like 99% sure the only difference between the 30B model and the big model is number of MP shards, learning rate, and batch size. All the other architectural hyperparams were held constant.

Does your prompt have a trailing space by chance?

**Comment by zhisbug** [comment]
TL;DR: 
With substantial help from @anj-s and Mingzhe's, we concluded that one of the original shards distributed by Meta is corrupted.

Details:
The shard `checkpoint_last-model_part-5-shard51.pt` is corrupted at the param key `decoder.layers.15.flat_param_0` at  indices[1241838:1241838+18] .

if you do the following:
```python
shard_path = "checkpoint_last-model_part-5-shard51.pt"
shard = torch_load_cpu(shard_path)
print(shard["model"]["decoder.layers.15.flat_param_0"][1241838:1241838+18])
```
It will output:
```
tensor([ 5.2691e-05,  4.5000e+03,  1.6427e-04,  5.5080e+03,  5.2256e+04,
         6.5880e+03,  3.0224e+04,  5.2631e-05,  7.5640e+03,  6.4969e-06,
         1.5259e-05,  1.5259e-05,  0.0000e+00,  5.0783e-04,  0.0000e+00,
        -0.0000e+00,  0.0000e+00, -4.8828e-03], dtype=torch.float16)
```
These 18 values are corrupted. As you can quickly tell, some values go to +1000s, which is very unusual for GPT weights.

It was a long story how we discover these 18 values, but essentially these 18 weights are the only 18 values that cause the difference between my consolidated version and the ground truth version (we checked MD5 and did some search).

They correspond to the fc1 weights in the 16th transformer layer (which is a critical layer compared to bias terms), and their magnitudes are substantial, I think they should be the reason that drives the model to produce collapsed outputs in my experiments.

@stephenroller Could you double check and see if our conclusion is correct? If yes -- could you update this shard, thank you!

**Comment by stephenroller** [comment]
@suchenzang Uhhh? No one else has had corrupted shards right? @patrickvonplaten, your copy is fine?

Sounds like we need to distribute md5sums.

**Comment by suchenzang** [comment]
Ack. Looking.

**Comment by suchenzang** [comment]
@stephenroller This is what I'm seeing:
```
(base) susanz@<old azure cluster>:/data/<redacted>/models/<redacted>/175B/raw$ md5sum checkpoint_last-model_part-5-shard51.pt
2af6ac9e2c440a29ad9ddab99af9a968  checkpoint_last-model_part-5-shard51.pt
```
Raw from S3:
```
(base) √ ~ % aws s3 cp s3://<redacted>/checkpoint_last-model_part-5-shard51.pt .
download: s3://<redacted>/checkpoint_last-model_part-5-shard51.pt to ./checkpoint_last-model_part-5-shard51.pt
(base) √ ~ % md5sum checkpoint_last-model_part-5-shard51.pt
2af6ac9e2c440a29ad9ddab99af9a968  checkpoint_last-model_part-5-shard51.pt
```
And from a presigned URL (backed by S3 ^):
```
(base) ?1 ~ % wget -O "checkpoint_last-model_part-5-shard51.pt" "https://d3q5sbpjbmb5k9.cloudfront.net/175b/checkpoint_last_20220503/checkpoint_last-model_part-5-shard51.pt?&<redacted>"
--2022-06-04 16:53:27--  <redacted>
Resolving d3q5sbpjbmb5k9.cloudfront.net (d3q5sbpjbmb5k9.cloudfront.net)... 99.86.105.5, 99.86.105.91, 99.86.105.38, ...
Connecting to d3q5sbpjbmb5k9.cloudfront.net (d3q5sbpjbmb5k9.cloudfront.net)|99.86.105.5|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 353677779 (337M) [binary/octet-stream]
Saving to: ‘checkpoint_last-model_part-5-shard51.pt’

checkpoint_last-model_part-5-shard51.pt  100%[================================================================================>] 337.29M  6.81MB/s    in 32s

2022-06-04 16:54:00 (10.5 MB/s) - ‘checkpoint_last-model_part-5-shard51.pt’ saved [353677779/353677779]
(base) √ ~ % md5sum checkpoint_last-model_part-5-shard51.pt
2af6ac9e2c440a29ad9ddab99af9a968  checkpoint_last-model_part-5-shard51.pt
```

This assumes our internal resharded checkpoints passed in `/data/<redacted>/models/<redacted>/175B/raw` from our old Azure cluster.  From the looks here - there isn't a file corruption issue wrt `checkpoint_last-model_part-5-shard51.pt`.  @zhisbug Can you confirm the md5sum you see for your `checkpoint_last-model_part-5-shard51.pt` file?

Will release md5sums for all the other files too: https://github.com/facebookresearch/metaseq/issues/137

**Comment by zhisbug** [comment]
@stephenroller @suchenzang 

Here is what I got:

```
$ md5sum 175B/checkpoint_last-model_part-5-shard51.pt
b2f3995bed96db012a6046f68fce4fe5  175B/checkpoint_last-model_part-5-shard51.pt
```
Seems it does not match the official MD5.

My download link has expired -- is there a way I can re-download this file? Thank you!

**Comment by suchenzang** [comment]
@zhisbug Please reply to the email from where you received the original link with the above information - you will receive a fresh link afterwards.

**Comment by zhisbug** [comment]
@suchenzang Just replied!

**Comment by suchenzang** [comment]
Link has be re-sent, closing for now - please open a new issue (and link to this one) if output still seems problematic!
